arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -0,0 +1,464 @@
+/*
+ * Copyright 2019 Broadcom
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <arch/arm/aarch64/cpu.h>
+#include <arch/arm/aarch64/arm_mmu.h>
+#include <linker/linker-defs.h>
+#include <sys/util.h>
+
+/* Set below flag to get debug prints */
+#define MMU_DEBUG_PRINTS	0
+/* To get prints from MMU driver, it has to initialized after console driver */
+#define MMU_DEBUG_PRIORITY	70
+
+#if MMU_DEBUG_PRINTS
+/* To dump page table entries while filling them, set DUMP_PTE macro */
+#define DUMP_PTE		0
+#define MMU_DEBUG(fmt, ...)	printk(fmt, ##__VA_ARGS__)
+#else
+#define MMU_DEBUG(...)
+#endif
+
+/* We support only 4kB translation granule */
+#define PAGE_SIZE_SHIFT		12U
+#define PAGE_SIZE		(1U << PAGE_SIZE_SHIFT)
+#define XLAT_TABLE_SIZE_SHIFT   PAGE_SIZE_SHIFT /* Size of one complete table */
+#define XLAT_TABLE_SIZE		(1U << XLAT_TABLE_SIZE_SHIFT)
+
+#define XLAT_TABLE_ENTRY_SIZE_SHIFT	3U /* Each table entry is 8 bytes */
+#define XLAT_TABLE_LEVEL_MAX	3U
```

Comment:
I'm a bit confused about the relation between `XLAT_TABLE_LEVEL_MAX` and `CONFIG_MAX_XLAT_TABLES`. Aren't those supposed to be the same value?

Reply:
`XLAT_TABLE_LEVEL_MAX `indicates maximum level of translation table in ARMv8-A, i.e. 3.
And `CONFIG_MAX_XLAT_TABLES `indicates total number of translation tables(minus base translation table).

So, `XLAT_TABLE_LEVEL_MAX `will remain same for all platforms.
However, `CONFIG_MAX_XLAT_TABLES `can change depending memory(normal+device) map of the SoC. That's why it is config option.